lib/controllers/issueish-detail-controller.js

```diff
@@ -44,14 +71,124 @@ export class BareIssueishDetailController extends React.Component {
       return <div>Issue/PR #{this.props.issueishNumber} not found</div>; // TODO: no PRs
     }
 
+    this.checkoutOp = this.nextCheckoutOp();
+
     return (
       <IssueishDetailView
         repository={repository}
         issueish={repository.issueish}
+        checkoutOp={this.checkoutOp}
         switchToIssueish={this.props.switchToIssueish}
       />
     );
   }
+
+  nextCheckoutOp() {
+    const {repository} = this.props;
+    const {issueish} = repository;
+
+    if (issueish.__typename !== 'PullRequest') {
+      return this.checkoutOp.disable(checkoutStates.HIDDEN, 'Cannot check out an issue');
+    }
+
+    if (this.props.isAbsent) {
+      return this.checkoutOp.disable(checkoutStates.HIDDEN, 'No repository found');
+    }
+
+    if (this.props.isLoading) {
+      return this.checkoutOp.disable(checkoutStates.DISABLED, 'Loading');
+    }
+
+    if (!this.props.isPresent) {
+      return this.checkoutOp.disable(checkoutStates.DISABLED, 'No repository found');
+    }
+
+    if (this.props.isMerging) {
+      return this.checkoutOp.disable(checkoutStates.DISABLED, 'Merge in progress');
+    }
+
+    if (this.props.isRebasing) {
+      return this.checkoutOp.disable(checkoutStates.DISABLED, 'Rebase in progress');
+    }
+
+    if (this.state.checkoutInProgress) {
+      return this.checkoutOp.disable(checkoutStates.DISABLED, 'Checking out...');
+    }
+
+    // Determine if we already have this PR checked out.
+
+    const headPush = this.props.branches.getHeadBranch().getPush();
+    const headRemote = this.props.remotes.withName(headPush.getRemoteName());
+
+    // (detect checkout from pull/### refspec)
+    const fromPullRefspec =
+      headRemote.getOwner() === repository.owner.login &&
+      headRemote.getRepo() === repository.name &&
+      headPush.getShortRemoteRef() === `pulls/${issueish.number}/head`;
+
+    // (detect checkout from head repository)
+    const fromHeadRepo =
+      headRemote.getOwner() === issueish.headRepository.owner.login &&
+      headRemote.getRepo() === issueish.headRepository.name &&
+      headPush.getShortRemoteRef() === issueish.headRefName;
+
+    if (fromPullRefspec || fromHeadRepo) {
+      return this.checkoutOp.disable(checkoutStates.CURRENT, 'Current');
+    }
+
+    return this.checkoutOp.enable();
+  }
+
+  async checkout() {
```

Comment:
@jcansdale, @grokys, @sguthals: I'd be curious to hear if I made different tradeoffs here than you did in Visual Studio, and what your reasoning was if so. I waffled between a few different implementations before settling on this one.

Essentially:

1. I'm fetching the PR ref from the head repository, not the magic `refs/pulls/123/head` ref that dotcom gives you.
2. I'm adding a new named remote for each unique head repository. The remote name is the owner of the head repository, so checking out a PR from `smashwilson/atom` to `atom/atom` will give you a remote called `smashwilson`. (If a remote called `smashwilson` already exists and doesn't point to that repo, it'll fail.)
3. If your local repo already has a branch (with any name) that tracks the head ref from the head repo remote, it'll be checked out (and pulled with `--ff-only`).
4. Otherwise, the new head ref is fetched and checked out as a branch called `pr-123/head-ref-name`.

My priorities were:

* To avoid creating more and more branches if you checkout a PR that you've already checked out.
* To have `git fetch` and `git pull` work as expected (meaning, update with new work from the head ref) while you have the PR branch checked out, even if you run them from another client.
* To let you see the hilarious branch names contributors or your teammates come up with :trollface:.

The big downside to the way I've written it is that you'll slowly accumulate junk remotes in your config, especially if you're an open-source maintainer and you test PRs from different contributors often. But I don't think there's a way to automatically clean them up, and if you fetch from the head URL directly you can't set up a tracking relationship.

Reply:
Yes, they all sound like a good ideas. We made a few mistakes here (which I'm not going to list :P) and we need to re-do out checking out of PRs. This sounds like a more sensible approach, so if it looks like it works for you then we should move to doing this I think.

Do you agree @jcansdale?

> The big downside to the way I've written it is that you'll slowly accumulate junk remotes in your config

The alternative to this is instead of creating named remotes, put the remote URL in the branch config:

```
[branch "pr-123/foo"]  
    remote = https://github.com/smashwilson/atom
```

However we couldn't do this as libgit2sharp doesn't support it. If you're shelling out to `git` this should work.

Reply:
> [..] put the remote URL in the branch config [..]

Ahhh ha! Thank you, this is what I was looking for 🙇I _thought_ this was possible, but I kept looking for ways to do it with the `git checkout -b` or `git branch` commands rather than just setting the config keys myself. (`--track` verifies that its argument corresponds to an actual remote.)

Reply:
@smashwilson,

> Otherwise, the new head ref is fetched and checked out as a branch called `pr-123/head-ref-name`.

The problem with this scheme (and the one we use in GHfVS) is that there's no indication that the branch is tracking a remote for a different fork.

I'd be inclined to use something along the lines of: `pr-123/owner/head-ref-name`. That way when one of my PRs is checked out from parent repository, the branch would be named something like this: `pr-123/jcansdale/fixes/666-nasty-bug` (where `fixes/666-nasty-bug` is my branch name). That would make it easy to see at a glance where branches are coming from.

Do you prefer `pr-123` over `pr/123`?

In GHfVS we use `pr/123-pull-request-title` (title not branch name). 😆

Reply:
@grokys, 

> Yes, they all sound like a good ideas. We made a few mistakes here (which I'm not going to list :P) and we need to re-do our checking out of PRs. This sounds like a more sensible approach, so if it looks like it works for you then we should move to doing this I think.

Pretty much. I think one of our mistakes was not including the owner name. A lot of people don't realize that PR branches are literally branches from another fork. We should make this more obvious.

> If your local repo already has a branch (with any name) that tracks the head ref from the head repo remote, it'll be checked out (and pulled with --ff-only).

Yup, we should be doing this.

Reply:
> The problem with this scheme (and the one we use in GHfVS) is that there's no indication that the branch is tracking a remote for a different fork.

I don't think that's particularly a problem, personally. The real thing we did wrong there IMO was trying to convert the PR title into a branch name, which results in stupidly long branch names at times.

Reply:
> I'd be inclined to use something along the lines of: pr-123/owner/head-ref-name. That way when one of my PRs is checked out from parent repository, the branch would be named something like this: pr-123/jcansdale/fixes/666-nasty-bug (where fixes/666-nasty-bug is my branch name). That would make it easy to see at a glance where branches are coming from.

👍 I really like this. There's a nice symmetry with the way git names remote tracking branches there, too.

> I don't think that's particularly a problem, personally.

Maybe not! That's likely something that we could surface with the appropriate UXR, maybe? In any case, I don't think it'd _hurt_.

> Do you prefer pr-123 over pr/123?

Kind of? I _think_ that one's a purely aesthetic concern though. There might be implications for ease of scripting against them but I feel like anything you write there is already going to be using a bunch of `sed`-fu either way.

Reply:
> In any case, I don't think it'd hurt.

Yeah, I probably worded this badly: what I should have said is perhaps "I don't think this is the biggest problem with our naming scheme" - the big mistake we made was turning the PR title into the branch name.

Reply:
> 👍 I really like this. There's a nice symmetry with the way git names remote tracking branches there, too.

I'm all about the symmetry. ⚛️ 

That's why I was also wondering about:
```
pulls/123-jcansdale/fixes/666-nasty-bug
```

That would keep my `heads` folder nice and organized.

![image](https://user-images.githubusercontent.com/11719160/42577969-6f71c694-851d-11e8-9b49-9bcd721fc1f8.png)

That way `remotes`, `tags`, `pulls` and `fixes` would all describe the elements in the contained namespace (if that makes sense).